service/datastore/client-rest/src/main/java/org/eclipse/kapua/service/datastore/client/rest/RestDatastoreClient.java

```diff
@@ -0,0 +1,641 @@
+        logger.debug("Query - converted query: '{}'", queryMap.toString());
+        long totalCount = 0;
+        Response queryResponse = null;
+        ArrayNode resultsNode = null;
+        try {
+            queryResponse = esClientProvider.getClient().performRequest(
+                    GET_ACTION,
+                    getSearchPath(typeDescriptor),
+                    Collections.<String, String>emptyMap(),
+                    EntityBuilder.create().setText(MAPPER.writeValueAsString(queryMap)).build(),
+                    new BasicHeader("Content-Type", ContentType.APPLICATION_JSON.toString()));
+            if (isRequestSuccessful(queryResponse)) {
+                JsonNode responseNode = MAPPER.readTree(EntityUtils.toString(queryResponse.getEntity()));
+                JsonNode hitsNode = responseNode.get(KEY_HITS);
+                totalCount = ((NumericNode) hitsNode.get(KEY_TOTAL)).asInt();
+                if (totalCount > Integer.MAX_VALUE) {
+                    throw new ClientException(ClientErrorCodes.ACTION_ERROR, "Total hits exceeds integer max value");
+                }
+                resultsNode = ((ArrayNode) hitsNode.get(KEY_HITS));
+            } else {
+                throw new ClientException(ClientErrorCodes.ACTION_ERROR, queryResponse.getStatusLine().getReasonPhrase());
+            }
+        } catch (ResponseException re) {
+            handleResponseException(re, typeDescriptor.getIndex(), "QUERY");
+        } catch (JsonProcessingException e) {
+            throw new ClientException(ClientErrorCodes.ACTION_ERROR, e, e.getLocalizedMessage());
+        } catch (IOException e) {
+            throw new ClientException(ClientErrorCodes.ACTION_ERROR, e, e.getLocalizedMessage());
+        }
+        ResultList<T> resultList = new ResultList<T>(totalCount);
+        if (resultsNode != null && resultsNode.size() > 0) {
+            for (JsonNode result : resultsNode) {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> object = MAPPER.convertValue(result.get(KEY_SOURCE), Map.class);
+                String id = ((TextNode) result.get(KEY_DOC_ID)).asText();
+                String index = ((TextNode) result.get(KEY_DOC_INDEX)).asText();
+                String type = ((TextNode) result.get(KEY_DOC_TYPE)).asText();
+                object.put(ModelContext.TYPE_DESCRIPTOR_KEY, new TypeDescriptor(index, type));
+                object.put(ModelContext.DATASTORE_ID_KEY, id);
+                object.put(QueryConverter.QUERY_FETCH_STYLE_KEY, queryFetchStyle);
+                resultList.add(modelContext.unmarshal(clazz, object));
+            }
+        }
+        return resultList;
+    }
+
+    @Override
+    public long count(TypeDescriptor typeDescriptor, Object query) throws ClientException {
+        checkClient();
+        JsonNode queryMap = queryConverter.convertQuery(query);
+        logger.debug("Query - converted query: '{}'", queryMap.toString());
```

Comment:
Remove `.toString()` is already called `.toString()` for each parameter passed as argument.